### PR TITLE
[SourceManager] Expose max usage of source location space as a Statistic

### PR DESCRIPTION
**Pull request by ilya-biryukov**
We have been running into source location exhaustion recently and want to use the statistics to monitor the usage in various files to be able to anticipate where the next problem will happen.

I picked `Statistic` because it can be written into a structured JSON file and is easier to consume by further automation.

This commit does not change any existing per-source-manager metrics exposed via `SourceManager::PrintStats()`. This does create some redundancy, but I also expect to be non-controversial because it aligns with the intended use of `Statistic`.

### CODE REVIEW

**Comment by ilya-biryukov** [comment]
cc @aeubanks @jyknight

File: clang/lib/Basic/SourceManager.cpp [hunk 1]
```diff
@@ -46,6 +47,12 @@ using namespace clang;
 using namespace SrcMgr;
 using llvm::MemoryBuffer;
 
+#define DEBUG_TYPE "source-manager"
+
+STATISTIC(
+    MaxUsedSLocBytes,
+    "Maximum number of bytes used by source locations (both loaded and local)");
```
Comment (usx95) [inline clang/lib/Basic/SourceManager.cpp:54]: nit: It may also be helpful to add some background on why this is important to track.

**Review by usx95 — APPROVED** [review]
LGTM. 
Please wait a couple of days before landing to give a chance to other folks for reviewing.

File: clang/lib/Basic/SourceManager.cpp (hunk 1 above)
Comment (ilya-biryukov) [reply clang/lib/Basic/SourceManager.cpp]: Done. I added a comment because the description string is something that tends to rather look like a one-liner.

**Comment by ilya-biryukov** [comment]
> LGTM. Please wait a couple of days before landing to give a chance to other folks for reviewing.

What would be a rationale to hold it off? It's perfectly within the [policy](https://llvm.org/docs/CodeReview.html#must-code-be-reviewed-prior-to-being-committed) to submit patches like this without too many reviewers and instead resolve any questions post-commit.

I would be happy to revert or update if people raise questions, it just seems like too much wait time for a patch that does not change much.

**Comment by usx95** [comment]
This is a simple enough patch, so I agree we can deal with concerns post-commit as well. Let's go ahead with landing this then.

**Pull request merged**